lib/chef/provider/package/windows.rb

```diff
@@ -64,16 +62,36 @@ def package_provider
         end
 
         def installer_type
+          # Depending on the installer, we may need to examine installer_type or
+          # source attributes, or search for text strings in the installer file
+          # binary to determine the installer type for the user. Since the file
+          # must be on disk to do so, we have to make this choice in the provider.
           @installer_type ||= begin
             if @new_resource.installer_type
               @new_resource.installer_type
             else
-              file_extension = ::File.basename(@new_resource.source).split(".").last.downcase
+              basename = ::File.basename(@new_resource.source)
+              file_extension = basename.split(".").last.downcase
 
               if file_extension == "msi"
                 :msi
               else
-                raise ArgumentError, "Installer type for Windows Package '#{@new_resource.name}' not specified and cannot be determined from file extension '#{file_extension}'"
+                contents = ::Kernel.open(::File.expand_path(@new_resource.source), 'rb', &:read) # TODO: limit data read in
+                case contents
+                when /inno/i # Inno Setup
+                  :inno
+                when /wise/i # Wise InstallMaster
+                  :wise
+                when /nsis/i # Nullsoft Scriptable Install System
+                  :nsis
+                else
+                  # if file is named 'setup.exe' assume installshield
+                  if basename == 'setup.exe'
+                    :installshield
+                  else
+                    fail Chef::Exceptions::AttributeNotFound, "Installer type for Windows Package '#{@new_resource.name}' not specified and cannot be determined from file extension '#{file_extension}'"
```

Comment:
I wouldn't use AttributeNotFound here, as it's used for something else specific, but I seem to be in an echo chamber so maybe it's not important.